### PR TITLE
Edit Distance Example

### PR DESCRIPTION
**Pull request by aprasad36**
## Description

Get a polished version of edit distance. Note that we should have a separate PR for each example.

I think we should implement the 3 versions of edit distance in separate files, so that David can pick which one he wants to show for 270 on a year-by-year basis:
https://www.geeksforgeeks.org/edit-distance-dp-5/

## TODO

- [x] Implement the m x n version
- [x] Clean-up and add annotations
- ~~[x] Implement 2 x m version~~
- ~~[x] Implement 1d array version~~
- [x] Implement version with separate costs for add, remove, insert.

## Questions

I absolutely hated the spammy callback prints, so I made a debug mode (see dp_array for more details).
Should we keep the debug mode?
For some reason, things break when accessing the array like OPT[i][j]. Any idea why this is happening?

## Status

- [x] I have read the guidelines in [CONTRIBUTING.md](https://github.com/itsdawei/dynamically_programmed/blob/master/CONTRIBUTING.md)
- [x] I have formatted my code using `yapf`
- [x] I have linted my code with `pylint`
- [x] I have tested my code by running `pytest`
- [x] I have added a one-line description of my change to the changelog in `HISTORY.md`
- [x] This PR is ready to go

### CODE REVIEW

File: docs/examples/edit_distance.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
+# Edit Distance
+
+- Given two strings `str1` and `str2`, what is the minimum number of edits
+- to convert `str1` to `str2`?
```
Comment (itsdawei) [inline docs/examples/edit_distance.md:4]: Should be combined with the line above.

File: docs/examples/index.md [hunk 2]
```diff
@@ -10,3 +10,4 @@ Here are the current examples:
 - [Knapsack](knapsack.md)
 - [Matrix Traversal](matrix_traversal.md)
 - [Excavation](excavation.md)
```
Comment (itsdawei) [inline docs/examples/index.md:12]: Can you remove Matrix Traversal and Excavation from this list? I forgot to do it when I removed it from the navigation bar.

File: docs/examples/edit_distance.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
+# Edit Distance
+
+- Given two strings `str1` and `str2`, what is the minimum number of edits
+  to convert `str1` to `str2`?
+- Each edit costs 1, and is either an add, delete, or replace.
```
Comment (itsdawei) [inline docs/examples/edit_distance.md:5]: Let's use the terminology in this [lecture note](https://web.stanford.edu/class/cs124/lec/med.pdf) by Stanford. Also can you briefly explain what each of the operation means (with a little bit of formality)? e.g., it's not obvious to a student what does an add/insertion mean.

File: docs/examples/edit_distance.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+- Goal: find a minimum cost set of edits that converts `str1` to `str2`.
```
Comment (itsdawei) [inline docs/examples/edit_distance.md:6]: Maybe for now just remove the concept of "cost". And instead just focus on "minimize the number of edits". As we've discussed earlier, we would like to put the extensions of edit distance (e.g., 1d array and alpha beta costs) into the blog. I want this to be something that a student can read in isolation of David's lecture on edit distance and understand fully.

File: docs/examples/edit_distance.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+## Dynamic Programming Solution
+
+> A fully executable example can be found on our [GitHub](https://github.com/itsdawei/dpvis/tree/main/demos/edit_distance_2d.py).
+
+Let $OPT[i, j]$ denote the cheapest conversion between the first $i$ characters of `str1`
+and the first $j$ characters of `str2`.
+
+**BASE CASES: $i$ or $j$ is $0$.** At least one of our strings is empty,
```
Comment (itsdawei) [inline docs/examples/edit_distance.md:15]: For simplicity, can you break this down to two base cases (i=0 and j=0)? Thanks!

File: docs/examples/edit_distance.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+the last letter of `str2`. At this point, we have to either add the last letter of 
+`str2` to `str1`, delete the last letter of `str1`, or replace the last letter of 
+`str1` with the last letter of `str2`. Each option is an edit, so we must record
+$OPT[i, j] = 1 + min(OPT[i, j - 1], OPT[i - 1, j], OPT[i - 1, j - 1])$.
+
+## Visualization with `dpvis`
+
+We can visualize this with `dpvis` as follows:
+
+```python linenums="1"  hl_lines="1 17 60-62"
```
Comment (itsdawei) [inline docs/examples/edit_distance.md:33]: No need to highlight the lines. I did it in the other on to illustrate what changed between the implementations.

File: docs/examples/edit_distance.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    # Initialize an m+1 x n+1 array
+    OPT = DPArray((m + 1, n + 1), array_name="Edit Distance", dtype=int)
+
+    # Base cases: either str1 or str2 is empty
+    # Then we have to pay to remove/add the remaining letters
+    for i in range(m + 1):
+        OPT[i, 0] = i
+    for j in range(n + 1):
+        OPT[0, j] = j
+    OPT.annotate("Base cases: no remaining letters in str1 or str2.")
```
Comment (itsdawei) [inline docs/examples/edit_distance.md:58]: I don't think this annotation is very informative unless you already understand what is going on in the algorithm. I think maybe put the formula in line 55 and line 57 here.

File: docs/examples/edit_distance.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+    # Fill OPT[][] iteratively
+    for i in range(m + 1):
+        for j in range(n + 1):
+            # Base case: either string is empty and has already been handled.
+            if i == 0 or j == 0:
+                pass
```
Comment (itsdawei) [inline docs/examples/edit_distance.md:65]: Don't use pass.

File: docs/examples/edit_distance.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    for i in range(m + 1):
+        OPT[i, 0] = i
+    for j in range(n + 1):
+        OPT[0, j] = j
+    OPT.annotate("Base cases: no remaining letters in str1 or str2.")
+
+    # Fill OPT[][] iteratively
+    for i in range(m + 1):
+        for j in range(n + 1):
+            # Base case: either string is empty and has already been handled.
```
Comment (itsdawei) [inline docs/examples/edit_distance.md:63]: We could just start iterating from 1?

File: docs/examples/edit_distance.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+            # At this point the last characters are different, so consider
+            # each possible action and pick the cheapest.
+            else:
+                indices = [
+                    (i, j - 1),  # Insert
+                    (i - 1, j),  # Remove
+                    (i - 1, j - 1)  # Replace
+                ]
+                OPT[i, j] = 1 + OPT.min(indices=indices)
```
Comment (itsdawei) [inline docs/examples/edit_distance.md:82]: Small thing but let's put the 1 and the end of the formula

File: docs/examples/edit_distance.md [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    # Test Example
+    str1 = "sunday"
+    str2 = "saturday"
+
+    dp_array = edit_distance(str1, str2)
+
+    display(dp_array, row_labels="_" + str1, column_labels="_" + str2)
+```
+
+This is what you will see when you execute the above code.
```
Comment (itsdawei) [inline docs/examples/edit_distance.md:98]: Hmmm... I'm thinking if we even need this section. What is the goal of this tutorial?

**Review by itsdawei — APPROVED** [review]
Left some comments, but we don't have to resolve all of them before merging. A lot of them are comments on the writing. Just merge when you've read through them and feel good about it.

**Comment by itsdawei** [comment]
We should add 1D optimization and alpha beta optimization in a different PR.

File: docs/examples/edit_distance.md (hunk 9 above)
Comment (aprasad36) [reply docs/examples/edit_distance.md:63]: Will do

File: docs/examples/index.md (hunk 2 above)
Comment (aprasad36) [reply docs/examples/index.md:12]: Done.

File: docs/examples/edit_distance.md (hunk 4 above)
Comment (aprasad36) [reply docs/examples/edit_distance.md:6]: I think big David wanted something else, so we ended up going for that.

**Review by ramirodcv — APPROVED** [review]

**Pull request merged**